### DISCUSSION TITLE
Orbit around selected item?

### DISCUSSION THREAD

**Discussion by seltzdesign** [microdee/OpenUnrealSpaceMouse#98] [Ideas]
I have to once again reference the SpaceMouse implementation in Rhino, which in our experience over the last 7 years or so with the SpaceMouse is the best implementation of their own driver we have come across.

One thing they do really well in Orbit Mode (which they call Object Mode) is that as soon as you select anything in the viewport it becomes the center of rotation for the SpaceMouse. That means that orbiting around a selected object and subsequently zooming into that object is super seemless, because it already has the right center of rotation. It's so natural in fact, that when the SpaceMouse behaves differently it is somewhat irritating.

Is there any chance we can have the center of rotation always be the object that is selected and only revert to the center of the screen when nothing is selected? Of course it should be an option.

Or did I just miss something and that is already possible?

Thanks for considering!

**Comment by microdee** [comment]
Yeah I thought of that as well originally when I added the orbitting behavior. The problem is that Unreal deals with entire worlds where your selection can be your skybox or your entire terrain, in which case you don't want to force your rotation around these kilometers sized objects. When working with terrain especially you always want to focus on the area you're working on, which might be hundreds of meters away from the center of its representing actor. That's why I ended up picking the surface of the object at the center of the screen at the beginning of movement for orbiting.

I guess it can be improved based on the screen coverage of the object's bounding box, like it will use object center if we're outside of the object's bounding box and that fits entirely on the screen, otherwise start using the point on surface method. Or we can find some volumetric algorithm to find a perfectly comfortable orbiting pivot inside the portion of the object visible on the screen but that's too much effort imo for how much time I have for this plugin right now.

Btw at static mesh editor viewports or material editor viewports the Rhino-like behavior would make more sense, however there's no universal way to express "selection" among all those 3D editor types.
